### PR TITLE
stats: flush stats even server is stuck initializing

### PR DESCRIPTION
**Pull request by ramaraochavali**
Signed-off-by: Rama Chavali <rama.rao@salesforce.com>
Description: When Envoy is stuck at initialization waiting for resources like cluster endpoints from management server, it does not set server stats and flush stats. It would be useful to update server stats and keep flushing the stats. A variant of issue described at https://github.com/envoyproxy/envoy/issues/4420

Risk Level: Low
Testing: Added automated test and also tested manually.
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by ramaraochavali** [comment]
The issue here is that `runOnAllThreads` completion callback is not called because workers are registered but not started and hence `mergeInternal` below  is not called to finish the merging process  and stats flush timer is never enabled again after the first flush.

https://github.com/envoyproxy/envoy/blob/54e22406c38987d5b0401f18b614501bc184f41f/source/common/stats/thread_local_store.cc#L170

File: source/server/server.cc [hunk 1]
```diff
@@ -166,27 +166,42 @@ void InstanceUtil::flushMetricsToSinks(const std::list<Stats::SinkPtr>& sinks,
...
-    server_stats_->days_until_first_cert_expiring_.set(
-        sslContextManager().daysUntilFirstCertExpires());
-    InstanceUtil::flushMetricsToSinks(config_.statsSinks(), stats_store_);
-    // TODO(ramaraochavali): consider adding different flush interval for histograms.
+  // If Envoy is not fully initialized, workers will not be started and mergeHistograms
+  // completion callback is not called immediately. As a result of this server stats will
+  // not be updated and flushed to stat sinks. So skip mergeHistograms call if workers are
+  // not started yet.
+  if (initManager().state() == Init::Manager::State::Initialized) {
+    // Disable any pending stats timers.
```
Comment (mattklein123) [inline source/server/server.cc]: Sorry why is this required? When would the timer be active at this point?

Comment (ramaraochavali) [reply source/server/server.cc]: It could be active if stats timer has been enabled in the else block and before it fires, server finishes initialization. Very rare case, but I thought safe to add this call.

Comment (ramaraochavali) [reply source/server/server.cc]: Sorry. Ignore my comment. You are right, it is not required. Removed it now.

File: test/server/server_test.cc [hunk 2]
```diff
@@ -134,6 +134,93 @@ TEST_F(RunHelperTest, ShutdownBeforeInitManagerInit) {
...
+
+// Class creates minimally viable server instance for testing and the initManager managed by it is
+// stuck in Initializing.
+class InitializingServerInstanceImplTest
+    : public testing::TestWithParam<Network::Address::IpVersion> {
+protected:
+  InitializingServerInstanceImplTest() : version_(GetParam()) {}
+
+  void initialize(const std::string& bootstrap_path) {
+    if (bootstrap_path.empty()) {
```
Comment (mattklein123) [inline test/server/server_test.cc]: I think this case is not used in your new test? Can you remove it?

File: test/server/server_test.cc [hunk 3]
```diff
@@ -134,6 +134,93 @@ TEST_F(RunHelperTest, ShutdownBeforeInitManagerInit) {
...
+      options_.config_path_ = TestEnvironment::temporaryFileSubstitute(
+          "test/config/integration/server.json", {{"upstream_0", 0}, {"upstream_1", 0}}, version_);
+    } else {
+      options_.config_path_ = TestEnvironment::temporaryFileSubstitute(
+          bootstrap_path, {{"upstream_0", 0}, {"upstream_1", 0}}, version_);
+    }
+    thread_local_ = std::make_unique<ThreadLocal::InstanceImpl>();
+    if (process_object_ != nullptr) {
```
Comment (mattklein123) [inline test/server/server_test.cc]: It seems like a bunch of this is not needed for the test? Instead of copying it all, can you refactor in same way to share the code?

File: test/server/server_test.cc [hunk 4]
```diff
@@ -134,6 +134,93 @@ TEST_F(RunHelperTest, ShutdownBeforeInitManagerInit) {
...
+
+// Validates that server stats are created even if the server is stuck with initialization.
+TEST_P(InitializingServerInstanceImplTest, StatsFlushWhenServerIsStillInitializing) {
+  options_.hot_restart_epoch_ = 3;
+
+  EXPECT_NO_THROW(initialize("test/server/empty_bootstrap.yaml"));
+
+  // Validate that server is still in initializing state but stats are created and flushed.
+  EXPECT_EQ(Init::Manager::State::Initializing, server_->initManager().state());
+  EXPECT_EQ(1L, TestUtility::findGauge(stats_store_, "server.live")->value());
```
Comment (mattklein123) [inline test/server/server_test.cc]: I'm not sure that this actually tests the change? Wouldn't these stats exist in the store even modulo the change? It seems like maybe you need a custom sink to make sure you get the stats?

**Comment by tehranian** [comment]
@ramaraochavali - What happens to Envoy's memory if it is stuck in `INITIALIZING` and it can't flush stats? Does Envoy's memory usage increase, unbounded? :) 

Asking because:
* We have several hosts which have Envoy installed, but those hosts are still using our legacy service mesh solution. Therefore, Envoy is idle from a traffic forwarding perspective. It's still talking to xDS endpoints.
* The Envoy process on these hosts are stuck in `INITIALIZING` state for some reason.
* The memory usage by Envoy has been growing steadily to to point where Envoy process is consuming several GB of memory. The `INITIALIZING` state seems to be a lead. That how I found this PR :)

**Comment by ramaraochavali** [comment]
@tehranian 
> What happens to Envoy's memory if it is stuck in INITIALIZING and it can't flush stats? Does Envoy's memory usage increase, unbounded? :)

AFAICT, the fact that stats are not flushed when it is in `INITIALIZING` state  should not affect Envoys memory usage.

File: test/server/server_test.cc (hunk 4 above)
Comment (ramaraochavali) [reply test/server/server_test.cc]: Yeah. I actually meant to test `flushStats` directly without writing a custom sink. I have corrected the test now. PTAL.

File: test/server/server_test.cc (hunk 3 above)
Comment (ramaraochavali) [reply test/server/server_test.cc]: Yup. We can do that. Recent commit has that refactoring.

File: test/server/server_test.cc [hunk 5]
```diff
@@ -32,7 +33,6 @@ using testing::StrictMock;
 
 namespace Envoy {
 namespace Server {
-namespace {
```
Comment (ramaraochavali) [inline test/server/server_test.cc]: This removal is needed, without this the actual friend class is not working - probably if we want to retain this we may have qualify including anonymous namespace while declaring it as friend. I thought it is OK to remove this.

File: test/server/server_test.cc [hunk 6]
```diff
@@ -180,6 +221,8 @@ class ServerInstanceImplTest : public testing::TestWithParam<Network::Address::I
     EXPECT_TRUE(server_->api().fileSystem().fileExists("/dev/null"));
   }
 
+  void flushStats() { server_->flushStats(); }
```
Comment (mattklein123) [inline test/server/server_test.cc]: I understand why you did this, but instead of reaching into the class under test, manually calling flush, looking at logs, etc. it would be a much better test if you used either one of the existing stat sinks or a simple mock/custom one to test this functionality. Can you work on that? I don't think it's that hard. Normally I wouldn't be asking for this level of testing, but this is pretty important functionality and IMO we should do our best to have a good test.

Comment (ramaraochavali) [reply test/server/server_test.cc]: I thought the rest of the code is well tested with existing tests and the new flow needs to be tested  - that is why I went with that approach. Now I have added `CustomStatsSink`. PTAL

File: test/server/server_test.cc (hunk 5 above)
Comment (mattklein123) [reply test/server/server_test.cc]: I think you can revert this change now.

File: test/server/server_test.cc [hunk 7]
```diff
@@ -197,10 +237,74 @@ class ServerInstanceImplTest : public testing::TestWithParam<Network::Address::I
...
+    server_->run();
+    startup_handle = nullptr;
+    server_ = nullptr;
+    thread_local_ = nullptr;
+  });
+
+  started.WaitForNotification();
+
+  // Wait till stats are flushed to custom sink and validate its value.
+  while (TestUtility::findCounter(stats_store_, "stats.flushed") == nullptr ||
```
Comment (mattklein123) [inline test/server/server_test.cc]: Can you potentially just use `waitForCounterEq` or factor it in a way that it can be used here?

File: test/server/server_test.cc [hunk 8]
```diff
@@ -197,10 +237,74 @@ class ServerInstanceImplTest : public testing::TestWithParam<Network::Address::I
...
+/**
+ * Static registration for the custom sink factory. @see RegisterFactory.
+ */
+REGISTER_FACTORY(CustomStatsSinkFactory, Server::Configuration::StatsSinkFactory);
+
+// Validates that server stats are flushed even when server is stuck with initialization.
+TEST_P(ServerInstanceImplTest, StatsFlushWhenServerIsStillInitializing) {
+  absl::Notification started;
+
+  auto server_thread = Thread::threadFactoryForTest().createThread([&] {
```
Comment (mattklein123) [inline test/server/server_test.cc]: This is copied, can you factor this out into some type of `startupTestServer` function or something?

Comment (ramaraochavali) [reply test/server/server_test.cc]: done

File: test/server/server_test.cc (hunk 7 above)
Comment (ramaraochavali) [reply test/server/server_test.cc]: I thought about that. We can not use that directly but we can refactor. I was actually thinking of doing a follow-up PR to move all `waitForxxxx` methods to TestUtility so that they can be reused better. For now I will refactor watiForCounter and do a follow-up for others
